File: packages/flutter/core_sdk/test/client_test.dart
import 'package:test/test.dart';
import 'package:core_sdk/core_sdk.dart';

import 'test_server.dart';

void main() {
  test('request attaches Bearer token and a correlation ID', () async {
    String? gotAuth;
    String? gotCorrelation;
    final server = await TestServer.start((req) async {
      gotAuth = req.headers.value('authorization');
      gotCorrelation = req.headers.value('x-correlation-id');
      await respondJson(req, 200, {});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl, tokenSource: StaticTokenSource('test-token'));
    await client.request('GET', '/whatever');

    expect(gotAuth, 'Bearer test-token');
    expect(gotCorrelation, isNotNull);
    expect(gotCorrelation, isNotEmpty);
  });

  test('request propagates an explicit correlation ID', () async {
    String? got;
    final server = await TestServer.start((req) async {
      got = req.headers.value('x-correlation-id');
      await respondJson(req, 200, {});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl);
    await client.request('GET', '/x', correlationId: 'caller-supplied-id');

    expect(got, 'caller-supplied-id');
  });

  test('request decodes a successful JSON response', () async {
    final server = await TestServer.start((req) async {
      await respondJson(req, 200, {'name': 'core-platform'});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl);
    final out = await client.request<String>('GET', '/x', decode: (j) => j['name'] as String);

    expect(out, 'core-platform');
  });

  test('request decodes the real error envelope into an ApiError', () async {
    final server = await TestServer.start((req) async {
      await respondJson(req, 403, {'code': 'ACCESS_DENIED', 'message': 'not allowed', 'correlationId': 'abc-123'});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl);
    try {
      await client.request('GET', '/x');
      fail('expected an ApiError');
    } catch (err) {
      expect(ApiError.isCode(err, ErrorCodes.accessDenied), isTrue);
      final apiErr = err as ApiError;
      expect(apiErr.statusCode, 403);
      expect(apiErr.correlationId, 'abc-123');
    }
  });

  test('request sends a JSON body with Content-Type', () async {
    String? gotContentType;
    String? gotBody;
    final server = await TestServer.start((req) async {
      gotContentType = req.headers.value('content-type');
      gotBody = await readBody(req);
      await respondJson(req, 201, {});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl);
    await client.request('POST', '/x', body: {'name': 'acme'});

    expect(gotContentType, contains('application/json'));
    expect(gotBody, '{"name":"acme"}');
  });

  test('GET is retried on a transient status until it succeeds', () async {
    var calls = 0;
    final server = await TestServer.start((req) async {
      calls++;
      if (calls < 3) {
        await respondJson(req, 503, {'code': 'DEPENDENCY_FAILURE', 'message': 'try again'});
        return;
      }
      await respondJson(req, 200, {});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl, retryPolicy: const DefaultRetryPolicy(maxAttempts: 3, baseDelay: Duration.zero));
    await client.request('GET', '/x');

    expect(calls, 3);
  });

  test('POST is not retried by default (retries where safe means GET-only)', () async {
    var calls = 0;
    final server = await TestServer.start((req) async {
      calls++;
      await respondJson(req, 503, {'code': 'DEPENDENCY_FAILURE', 'message': 'down'});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl, retryPolicy: const DefaultRetryPolicy(maxAttempts: 3, baseDelay: Duration.zero));
    await expectLater(client.request('POST', '/x'), throwsA(isA<ApiError>()));

    expect(calls, 1);
  });

  test('a non-transient status (404) is not retried', () async {
    var calls = 0;
    final server = await TestServer.start((req) async {
      calls++;
      await respondJson(req, 404, {'code': 'RESOURCE_NOT_FOUND', 'message': 'missing'});
    });
    addTearDown(server.close);

    final client = CoreClient(server.baseUrl, retryPolicy: const DefaultRetryPolicy(maxAttempts: 3, baseDelay: Duration.zero));
    try {
      await client.request('GET', '/x');
      fail('expected an ApiError');
    } catch (err) {
      expect(ApiError.isCode(err, ErrorCodes.notFound), isTrue);
    }
    expect(calls, 1);
  });
}

File: packages/typescript/core-sdk/src/client.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import { CoreClient } from "./client.js";
import { ApiError } from "./errors.js";
import { StaticTokenSource } from "./auth.js";
import { DefaultRetryPolicy } from "./retry.js";

async function withServer(handler: http.RequestListener, run: (baseUrl: string) => Promise<void>) {
  const server = http.createServer(handler);
  await new Promise<void>((resolve) => server.listen(0, resolve));
  const address = server.address();
  const port = typeof address === "object" && address ? address.port : 0;
  try {
    await run(`http://127.0.0.1:${port}`);
  } finally {
    await new Promise<void>((resolve) => server.close(() => resolve()));
  }
}

test("request attaches Bearer token and a correlation ID", async () => {
  let gotAuth = "";
  let gotCorrelation = "";
  await withServer(
    (req, res) => {
      gotAuth = req.headers["authorization"] ?? "";
      gotCorrelation = (req.headers["x-correlation-id"] as string) ?? "";
      res.writeHead(200).end();
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl, { tokenSource: new StaticTokenSource("test-token") });
      await client.request("GET", "/whatever");
      assert.equal(gotAuth, "Bearer test-token");
      assert.notEqual(gotCorrelation, "");
    },
  );
});

test("request propagates an explicit correlation ID", async () => {
  let got = "";
  await withServer(
    (req, res) => {
      got = (req.headers["x-correlation-id"] as string) ?? "";
      res.writeHead(200).end();
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl);
      await client.request("GET", "/x", undefined, { correlationId: "caller-supplied-id" });
      assert.equal(got, "caller-supplied-id");
    },
  );
});

test("request decodes a successful JSON response", async () => {
  await withServer(
    (_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" }).end(JSON.stringify({ name: "core-platform" }));
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl);
      const out = await client.request<{ name: string }>("GET", "/x");
      assert.equal(out.name, "core-platform");
    },
  );
});

test("request decodes the real error envelope into an ApiError", async () => {
  await withServer(
    (_req, res) => {
      res.writeHead(403, { "Content-Type": "application/json" }).end(
        JSON.stringify({ code: "ACCESS_DENIED", message: "not allowed", correlationId: "abc-123" }),
      );
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl);
      await assert.rejects(
        () => client.request("GET", "/x"),
        (err: unknown) => {
          assert.ok(ApiError.is(err, "ACCESS_DENIED"));
          assert.equal((err as ApiError).statusCode, 403);
          assert.equal((err as ApiError).correlationId, "abc-123");
          return true;
        },
      );
    },
  );
});

test("request sends a JSON body with Content-Type", async () => {
  let gotContentType = "";
  let gotBody = "";
  await withServer(
    (req, res) => {
      gotContentType = req.headers["content-type"] ?? "";
      let raw = "";
      req.on("data", (chunk) => (raw += chunk));
      req.on("end", () => {
        gotBody = raw;
        res.writeHead(201).end();
      });
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl);
      await client.request("POST", "/x", { name: "acme" });
      assert.equal(gotContentType, "application/json");
      assert.deepEqual(JSON.parse(gotBody), { name: "acme" });
    },
  );
});

test("GET is retried on a transient status until it succeeds", async () => {
  let calls = 0;
  await withServer(
    (_req, res) => {
      calls++;
      if (calls < 3) {
        res.writeHead(503, { "Content-Type": "application/json" }).end(JSON.stringify({ code: "DEPENDENCY_FAILURE", message: "try again" }));
        return;
      }
      res.writeHead(200).end();
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl, { retryPolicy: new DefaultRetryPolicy(3, 0) });
      await client.request("GET", "/x");
      assert.equal(calls, 3);
    },
  );
});

test("POST is not retried by default (retries where safe means GET-only)", async () => {
  let calls = 0;
  await withServer(
    (_req, res) => {
      calls++;
      res.writeHead(503, { "Content-Type": "application/json" }).end(JSON.stringify({ code: "DEPENDENCY_FAILURE", message: "down" }));
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl, { retryPolicy: new DefaultRetryPolicy(3, 0) });
      await assert.rejects(() => client.request("POST", "/x"));
      assert.equal(calls, 1);
    },
  );
});

test("a non-transient status (404) is not retried", async () => {
  let calls = 0;
  await withServer(
    (_req, res) => {
      calls++;
      res.writeHead(404, { "Content-Type": "application/json" }).end(JSON.stringify({ code: "RESOURCE_NOT_FOUND", message: "missing" }));
    },
    async (baseUrl) => {
      const client = new CoreClient(baseUrl, { retryPolicy: new DefaultRetryPolicy(3, 0) });
      await assert.rejects(
        () => client.request("GET", "/x"),
        (err: unknown) => ApiError.is(err, "RESOURCE_NOT_FOUND"),
      );
      assert.equal(calls, 1);
    },
  );
});
